modules/tinymce/src/themes/silver/test/ts/browser/editor/toolbar/ToolbarButtonContextTest.ts

```diff
@@ -0,0 +1,603 @@
+        const hook = TinyHooks.bddSetup<Editor>({
+          base_url: '/project/tinymce/js/tinymce',
+          toolbar: 't6',
+          statusbar: false,
+          setup: (ed: Editor) => {
+            registerMode(ed);
+
+            scenario.buttonSetupSix(ed);
+          }
+        }, [], true);
+
+        it(`TINY-11211: Toolbar ${scenario.label} buttonApi.setEnabled should overwrite the context`, async () => {
+          const editor = hook.editor();
+          editor.mode.set('design');
+          scenario.assertButtonDisabled('t6');
+
+          editor.mode.set('testmode');
+          scenario.assertButtonDisabled('t6');
+
+          editor.mode.set('readonly');
+          scenario.assertButtonDisabled('t6');
+
+          editor.mode.set('testmode');
+          scenario.assertButtonDisabled('t6');
+
+          editor.mode.set('design');
+          scenario.assertButtonEnabled('t6');
+          // Since the onSetup for this button is listening for the NodeChange event, manually dispatching a nodeChanged to trigger a state change
+          editor.nodeChanged();
+          scenario.assertButtonDisabled('t6');
+        });
+      });
+
+      context('Toolbar button spec with context: mode:readonly and onSetup buttonApi setEnabled(true)', () => {
+        const hook = TinyHooks.bddSetup<Editor>({
+          base_url: '/project/tinymce/js/tinymce',
+          toolbar: 't7',
+          statusbar: false,
+          setup: (ed: Editor) => {
+            registerMode(ed);
+
+            scenario.buttonSetupSeven(ed);
+          }
+        }, [], true);
+
+        it(`TINY-11211: Toolbar ${scenario.label} buttonApi.setEnabled should overwrite the context`, async () => {
+          const editor = hook.editor();
+          editor.mode.set('design');
+          scenario.assertButtonEnabled('t7');
+
+          editor.mode.set('testmode');
+          scenario.assertButtonDisabled('t7');
+
+          editor.mode.set('readonly');
+          scenario.assertButtonEnabled('t7');
+
+          editor.mode.set('testmode');
+          scenario.assertButtonDisabled('t7');
+
+          editor.mode.set('design');
+          scenario.assertButtonDisabled('t7');
+          editor.nodeChanged();
+          scenario.assertButtonEnabled('t7');
+        });
+      });
+
+      context('Toolbar button spec with context: mode:design and onSetup buttonApi setEnabled(false), without NodeChange', () => {
+        const hook = TinyHooks.bddSetup<Editor>({
+          base_url: '/project/tinymce/js/tinymce',
+          toolbar: 't8',
+          statusbar: false,
+          setup: (ed: Editor) => {
+            registerMode(ed);
+
+            scenario.buttonSetupEight(ed);
+          }
+        }, [], true);
+
+        it(`TINY-11211: Toolbar ${scenario.label} should be enabled in uiEnabled mode`, async () => {
+          const editor = hook.editor();
+          editor.mode.set('design');
+          scenario.assertButtonEnabled('t8');
+
```

Comment:
For this button configuration without a `NodeChange` or any other events, just `setEnabled(false)`. I'm not sure if we need to care about this? With this change, we always run on `NodeChange` or `SwitchMode`, however the `setEnabled` is only called once, so this button gets enabled. 
```
    ed.ui.registry.addButton('t8', {
      icon: 'italic',
      text: 'Test Menu Item 8',
      context: 'mode:design',
      onAction: Fun.noop,
      onSetup: (api) => {
        api.setEnabled(false);
        return Fun.noop;
      }
    });
```

Reply:
The current behaviour would enable a button that is disabled when the mode is switched to design, and disable a button that is enabled when the mode is switched to readonly. We would expect that any button that needs to be enabled or disabled in a specific mode would be setup to trigger the api on the 'SwitchMode' event. Supporting anything else would be unexpected. With the new context API, that workaround is solved as well 👍